### DISCUSSION TITLE
统一rop表达

### DISCUSSION THREAD

**Discussion by qiufuyu123** [Physics365/991CN-X-CW-Decompilation#42] [Ideas]
@Xyzstk 能不能把你的rop .bin 文件以汇编的形式写一下？
类似于这个：
![image](https://github.com/Physics365/991CN-X-CW-Decompilation/assets/60765480/0926fb61-db8f-45ad-b8e3-d7424770c184)
这种汇编可以直接通过 [汇编器](https://github.com/Physics365/991CN-X-CW-Decompilation/blob/main/ROP/ROPCodes/asm.py) 编译  
通过汇编的方式写rop大家更能理解一下具体的执行逻辑，也可以进一步统一ropstring的格式  
[汇编器语法参考](https://github.com/Physics365/991CN-X-CW-Decompilation/blob/main/ROP/ROPCodes/Tutor.md)
😃

**Comment by EwTE-H** [comment]
不可行。

机器码可以和指令一一对应，但是rop并非指令的形式。尤其是当数据互相重叠的时候（你可以看一眼我写的那个内存编辑器。大量数据重叠）而且，哪怕没有重叠，用图示的方法也不好，因为指令可以对应机器码，而rop不能对应。可以要求作者以某种方式解释代码，但不能强求统一格式。

我自己的草稿上用这样的格式：
```
start: D59A < D1B8
20D72; [out], //er14
2237A; #

out: D2DA < (<<branch2+E);

loop: D8EC < D1C4
0C06C; [ptr] //er8
18812; ptr_copied <, //er2
08F28; [to_copy], , //er8
08F28; , ,
f_221AE;
f_08706;
121A8; [key] //er0
f_0F748;
08FC2; key //er2
15C48; , , , , , [midjmp], //er14
0C06C; [key2] //er8
08F28; [key3], BFFD, //er8, er10
08F26; , key3, //er10
08FC2; 482C, //er2
23332; #
```

**Reply by user202729** [reply]
If overlapping instructions is the whole issue, then can't you just improve the assembler to handle that?

Overall I find writing ROP by hand is mostly being the manual compiler anyway. Given the assembly format supports `hex` instruction as an escape if the assembler can't automatically handle it anyway, I'd recommend letting the assembler doing as much work as possible, and improving the assembler when it is insufficient.

(is @EwTE-H the same as @Xyzstk ?)

**Reply by qiufuyu123** [reply]
yeah, basically my assembler supports `hex` instruction
BTW, @EwTE-H is NOT same as @Xyzstk

**Comment by EwTE-H** [comment]
语法上有几个可以提升的地方
- sym.txt只包含少数符号。最好能允许程序员**在代码内**自己定义符号。（改sym.txt就像改编译器头文件一样）
- 有循环的程序经常需要进行复制，最常见的就是复制代码重建栈。如果还包括复制数据那就更烧脑了，因为数据的流向经常会很混乱。语法中最好能提供一些支持。
- 数据/代码重叠问题。手写重叠代码也挺坐牢的。能否在语法中设置负的space，从而支持代码重叠？（负的space代表倒退一定数目的字节）
- 关键字较多，符号较少，看起来不够美观。

**Reply by qiufuyu123** [reply]
- sym.txt允许自定义符号，可以按照sym.txt内的格式增加
- 这个目前的实现方式是memcpy+jump, 可以参考loop.asm
![image](https://github.com/Physics365/991CN-X-CW-Decompilation/assets/60765480/085981e6-375e-4338-8ecc-a6be81e8f9cc)
- 负的space是啥意思啊，我一般通过hex xxxxxx( 一些占位字符实现重叠) 如果你想要设置程序计算的基地址，可以使用#org xxx
![image](https://github.com/Physics365/991CN-X-CW-Decompilation/assets/60765480/b1262f60-b6db-4bd6-a3ce-a172e4ac8a52)
- 这个确实，还没想好怎么弄

**Reply by qiufuyu123** [reply]
复制数据为了保证灵活性  
比如重新建栈可能每个人设置的位置不同？  
比如数据放在栈上的时候，重新建栈如果不想让数据被覆盖，可能需要准备两个缓冲区，然后用 adr xxx 做偏移计算出原来栈上的数据之类的  
所以就没有用关键字实现，只能用memcpy, jpop之类的方式

**Comment by EwTE-H** [comment]
尝试yy一点语法

# 汇编器语法

## 1. #org  
开启一个新的代码段并设置起始地址。本段中所有的地址（标签计算）以这个地址开始。也可以使用标签定义地址。  
```
#org 3a3b ; 这一段的起始地址为3a3b
hex .. ..
label: ; 标签位置在3a3c
#org 4a4b ; 这一段的起始地址为4a4b
#org &label+2 ; 这一段的起始位置为3a3e
#org &+10 ; 这一段的起始位置为3a4e
```
#org后跟大括号代表插入代码段，大括号结束时回到之前的段。    
```
#org d522
hex .. .. ; 这里是d524
#org &-8 { ; 这一段的起始位置为d51c
    hex .. .. .. ..
}
; 这里又回到了d524
```
如果两段重合，编译器会合并代码，一旦出现冲突就会报错。
## 2. hex  
```
hex 3. .. 30 43
```
直接写16进制，内容会直接原封不动视为一串16进制。.代表可以取任何值（由编译器决定）。hex可以省略。  

## 3. 标签/取地址  

```
#org d522
hex 33 33 
loop:   ;<--此时，标签代表的地址是(0xd522+2=0xd524)
hex .. .. .. ..
adr loop ; <-- 'adr loop' 会被替换成：hex 24 d5 (d524, loop的地址)
adr loop 55 ; <- 0xd524+55
adr loop -2 ; <- 0xd522
adr +2 ; <-- 不写标签名，则表示从当前地址(d52e)起算，替换成d530
&loop ; 这四条的含义分别与上面四条相同
&loop+55
&loop-2
&+2
```

## 4. 函数标签  
所有可用标签请查看sym.txt  
注意，所有标签使用需要单独成行
```
例子：sym.txt 115 行：
20840       and r1, 15, sll r0, 4, or r1, r0
当你在汇编中，使用
and r1, 15, sll r0, 4, or r1, r0
会被替换为 hex 40 08 .2 ..
```

```
例子2：sym.txt 115 行：
@0F746       read_key
当你在汇编中，使用
read_key
会被替换为 hex 48 f7 .2 ..(注意,当地址前面带'@',汇编器会自动将地址+2)
```

用#def来定义标签，如`#def @22040 diag`。

带参数个数的函数标签：在地址后面跟括号，里面写调用这个标签时pop的字节数。pop pc不算在内。例如pop qr0的标签可以写作`#def 130e2(8) pop qr0`。在程序中调用这类函数标签，后面必须跟括号，里面是相应数目的字节。例如`pop qr0 (31 32 .. .. .. .6 7. fe)`。

## 5.使用asm.py
```
python3.exe .\asm.py xxx (你想编译的文件）
```

**Reply by qiufuyu123** [reply]
### 新编译器已经全部支持！！

[语法指南](https://github.com/Physics365/991CN-X-CW-Decompilation/blob/main/ROP/ROPCodes/Tutor.md)
（语法做了一些微调，其他基本一样）

**Reply by qiufuyu123** [reply]
[百度贴吧的演示](https://tieba.baidu.com/p/9047618339?pid=150410469697&cid=0#150410469697)
